### PR TITLE
feat: respect the npmrc

### PR DESCRIPTION
**Pull request by Spencer17x**
Related to https://github.com/pnpm/pnpm/pull/4422

### CODE REVIEW

File: packages/plugin-commands-init/package.json [hunk 1]
```diff
@@ -40,6 +40,8 @@
     "@pnpm/config": "workspace:13.13.2",
     "@pnpm/error": "workspace:2.1.0",
     "@pnpm/write-project-manifest": "workspace:2.0.11",
+    "camelcase-keys": "^6.2.2",
+    "read-ini-file": "^3.1.0",
```
Comment (zkochan) [inline packages/plugin-commands-init/package.json]: all the settings from `.npmrc` are already passed to the plugin via the `rawConfig` option.

Comment (spencer17x) [reply packages/plugin-commands-init/package.json]: I have modified it, please check whether it is correct @zkochan

File: packages/plugin-commands-init/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  }
+
+  const author = unParsePerson(camelcaseKeys(authorInfo))
+  if (author) packageJson.author = author
+  return camelcaseKeys(packageJson)
+}
+
+export async function parseLocalConfig (prefix: string): Promise<Record<string, string>> {
+  const { config: { rawConfig } } = await getConfigs({
```
Comment (zkochan) [inline packages/plugin-commands-init/src/utils.ts]: rawConfig is already passed to the plugin. You just need to add the type to opts.

**Review by zkochan — CHANGES_REQUESTED** [review]
add tests

**Comment by Spencer17x** [comment]
@zkochan I added the unit test and changed the way rawConfig was read

File: packages/plugin-commands-init/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  const name = person.name ?? ''
+  const u = person.url ?? person.web
+  const url = u ? (' (' + u + ')') : ''
+  const e = person.email ?? person.mail
+  const email = e ? (' <' + e + '>') : ''
+  return name + email + url
+}
+
+function workWithInitModule (localConfig: Record<string, string>) {
+  if ('initModule' in localConfig && localConfig.initModule) {
```
Comment (zkochan) [inline packages/plugin-commands-init/src/utils.ts]: init-module is not covered with a test.

File: packages/plugin-commands-init/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    const filePath = path.resolve(localConfig.initModule)
+    spawnSync('node', [filePath], {
+      stdio: 'inherit',
+    })
+    delete localConfig.initModule
+  }
+  return localConfig
+}
+
+function workWithIncludeWorkspaceRoot (localConfig: Record<string, string>) {
```
Comment (zkochan) [inline packages/plugin-commands-init/src/utils.ts]: if it is not implemented, no need to add this function.

**Comment by Spencer17x** [comment]
@zkochan I made some changes, please review it

File: packages/plugin-commands-init/test/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+})
+
+test('run the unParsePerson function', () => {
+  const person: Person = {
+    email: 'xxxxxx@pnpm.com',
+    name: 'pnpm',
+    url: 'https://www.github.com/pnpm',
+  }
+  const expectAuthor = 'pnpm <xxxxxx@pnpm.com> (https://www.github.com/pnpm)'
+  expect(unParsePerson(person)).toBe(expectAuthor)
```
Comment (BlackHole1) [inline packages/plugin-commands-init/test/utils.test.ts]: Miss `string` test.

`unParsePerson function params` type is `string | Person`

File: packages/plugin-commands-init/src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+export function unParsePerson (person: string | Person): string {
+  if (typeof person === 'string') {
+    return person
+  }
+  const name = person.name ?? ''
+  const u = person.url ?? person.web
+  const url = u ? (' (' + u + ')') : ''
+  const e = person.email ?? person.mail
+  const email = e ? (' <' + e + '>') : ''
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]:
```suggestion
  const url = u ? ` (${u})`
  const e = person.email ?? person.mail
  const email = e ? ` <${e}>` : ''
```

File: packages/plugin-commands-init/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+export interface Person {
+  name?: string
+  email?: string
+  url?: string
+  web?: string
+  mail?: string
+}
+
+export function unParsePerson (person: string | Person): string {
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]: personToString

File: packages/plugin-commands-init/src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  const e = person.email ?? person.mail
+  const email = e ? (' <' + e + '>') : ''
+  return name + email + url
+}
+
+export function workWithInitModule (localConfig: Record<string, string>) {
+  const { initModule, ...restConfig } = localConfig
+  if ('initModule' in localConfig) {
+    const filePath = path.resolve(localConfig.initModule)
+    const isFileExist = fs.existsSync(filePath)
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts:27]:
```suggestion
    if (['.js', '.cjs'].include(path.extname(filePath)) && fs.existsSync(filePath)) {
      spawnSync('node', [filePath], {
        stdio: 'inherit',
      })
    }
    const isFileExist = fs.existsSync(filePath)
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by Spencer17x** [comment]
@BlackHole1 I have modified it, please check it

File: packages/plugin-commands-init/src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+export interface Person {
+  name?: string
+  email?: string
+  url?: string
+  web?: string
+  mail?: string
+}
+
+export function personToString (person: string | Person): string {
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]:
```suggestion
export function personToString (person: Person): string {
```

The string type is not used in the existing code

File: packages/plugin-commands-init/src/utils.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    const filePath = path.resolve(localConfig.initModule)
+    const isFileExist = fs.existsSync(filePath)
+    if (['.js', '.cjs'].includes(path.extname(filePath)) && isFileExist) {
+      spawnSync('node', [filePath], {
+        stdio: 'inherit',
+      })
+    }
+  }
+  return restConfig
+}
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]:
```suggestion
  if (initModule) {
    const filePath = path.resolve(initModule)
    const isFileExist = fs.existsSync(filePath)

    if (['.js', '.cjs'].includes(path.extname(filePath)) && isFileExist) {
       spawnSync('node', [filePath], {
        stdio: 'inherit',
      })
    }
  }
  
  return restConfig;
}
```

File: packages/plugin-commands-init/src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    if (isInitKey) {
+      const key = localConfigKey.replace('init', '')
+      const isAuthorKey = key.startsWith('Author')
+      if (isAuthorKey) {
+        authorInfo[key.replace('Author', '')] = value
+      } else {
+        packageJson[key] = value
+      }
+    }
+  }
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]:
```suggestion
  const packageJson: Record<string, string> = {}
  const authorInfo: Record<string, string> = {}

  for (const localConfigKey in localConfig) {
    if (localConfigKey.startsWith('init')) {
      const pureKey = localConfigKey.replace('init', '')
      const value = localConfig[localConfigKey]

      if (pureKey.startsWith('Author')) {
        authorInfo[key.replace('Author', '')] = value
      } else {
        packageJson[key] = value
      }
    }
  }
```

File: packages/plugin-commands-init/src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+  const author = personToString(camelcaseKeys(authorInfo))
+  if (author) packageJson.author = author
```
Comment (BlackHole1) [inline packages/plugin-commands-init/src/utils.ts]:
```suggestion
  if (author) {
    packageJson.author = author
  }
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by BlackHole1** [comment]
nit: Appropriate blank lines can help with code reading.

**Comment by Spencer17x** [comment]
@BlackHole1 I made the changes

**Review by BlackHole1 — APPROVED** [review]
LGTM

File: packages/plugin-commands-init/test/utils.test.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import { workWithInitModule, personToString } from '@pnpm/plugin-commands-init/lib/utils'
+
+test('run the workWithInitModule function', () => {
+  const rawConfig = {
+    initVersion: '2.0.0',
+    initModule: '~/.pnpm-init.js',
```
Comment (zkochan) [inline packages/plugin-commands-init/test/utils.test.ts]: how do you verify that this module was executed?

Comment (spencer17x) [reply packages/plugin-commands-init/test/utils.test.ts]: I have made a supplement, please review it

File: packages/plugin-commands-init/test/utils.test.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+import * as fs from 'fs'
+import * as path from 'path'
+
+test('run the workWithInitModule function', () => {
+  const tmpDir = tempDir()
+  const rawConfig = {
+    initVersion: '2.0.0',
+    initModule: path.join(tmpDir, 'init-module.js'),
+  }
+  fs.writeFileSync(rawConfig.initModule, `const fs = require('fs'); fs.writeFileSync('${path.join(tmpDir, 'test.txt')}', 'test');`)
```
Comment (BlackHole1) [inline packages/plugin-commands-init/test/utils.test.ts]: use `@pnpm/test-fixtures`

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by zkochan** [comment]
Tests are failing on Windows.

File: packages/plugin-commands-init/test/utils.test.ts (hunk 14 above)
Comment (spencer17x) [reply packages/plugin-commands-init/test/utils.test.ts]: I made the change, please review it

**Review by zkochan — APPROVED** [review]

**Pull request merged**